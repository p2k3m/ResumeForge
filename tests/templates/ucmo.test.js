import { parseContent } from '../../services/parseContent.js';
import Handlebars from '../../lib/handlebars.js';
import fs from 'fs/promises';
import path from 'path';
import puppeteer from 'puppeteer';
import zlib from 'zlib';

function tokensToHtml(tokens) {
  return tokens
    .map((t, i) => {
      const text = t.text || '';
      if (t.type === 'link') {
        const next = tokens[i + 1];
        const space = next && next.text && !/^\s/.test(next.text) ? ' ' : '';
        return `<a href="${t.href}">${text.trim()}</a>${space}`;
      }
      if (t.style === 'bolditalic') return `<strong><em>${text}</em></strong>`;
      if (t.style === 'bold') return `<strong>${text}</strong>`;
      if (t.style === 'italic') return `<em>${text}</em>`;
      if (t.type === 'newline') return '<br>';
      if (t.type === 'tab') return '<span class="tab"></span>';
      if (t.type === 'bullet') return '<span class="bullet">â€¢</span> ';
      return text;
    })
    .join('');
}

test('ucmo template renders with contact bar and logo', async () => {
  const input = 'Jane Doe\n# Experience\n- Built things\n# Education\n- BSc Stuff';
  const data = parseContent(input);
  const tplSrc = await fs.readFile(path.resolve('templates', 'ucmo.html'), 'utf8');
  const htmlData = {
    ...data,
    phone: '555-555-5555',
    email: 'jane@example.com',
    cityState: 'Warrensburg, MO',
    linkedin: 'linkedin.com/in/jane',
    sections: data.sections.map((sec) => ({
      ...sec,
      items: sec.items.map(tokensToHtml)
    }))
  };
  const html = Handlebars.compile(tplSrc)(htmlData);
  expect(html).toMatchSnapshot('html');
  let browser;
  try {
    browser = await puppeteer.launch({ args: ['--no-sandbox'] });
  } catch (err) {
    console.warn('Puppeteer launch failed, skipping PDF snapshot:', err.message);
    return;
  }
  const page = await browser.newPage();
  await page.setContent(html, { waitUntil: 'networkidle0' });
  const pdfBuffer = await page.pdf({ format: 'A4', printBackground: true });
  await browser.close();

  const pdfStr = pdfBuffer.toString('latin1');
  const objects = Object.fromEntries(
    pdfStr
      .split('endobj')
      .map((chunk) => {
        const id = chunk.match(/(\d+) 0 obj/);
        const base = chunk.match(/BaseFont \/([^\s]+)/);
        return id && base ? [id[1], base[1]] : null;
      })
      .filter(Boolean)
  );
  const fontMap = {};
  const fontRefs = [...pdfStr.matchAll(/\/F(\d+) (\d+) 0 R/g)];
  for (const [, id, obj] of fontRefs) {
    const base = objects[obj];
    if (base) fontMap[`F${id}`] = base;
  }

  const start = pdfBuffer.indexOf(Buffer.from('stream')) + 6;
  const nl = pdfBuffer.indexOf('\n', start) + 1;
  const end = pdfBuffer.indexOf(Buffer.from('endstream'), nl);
  let content = pdfBuffer.slice(nl, end);
  try {
    content = zlib.inflateSync(content).toString();
  } catch {
    content = content.toString();
  }

  expect({ fontMap, content }).toMatchSnapshot('pdf');
});
