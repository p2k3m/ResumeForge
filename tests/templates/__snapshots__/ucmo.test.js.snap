// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ucmo template renders with contact bar and logo: html 1`] = `
"<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8" />
  <title>Resume</title>
  <style>
    body { font-family: 'Times New Roman', serif; margin: 50px; color: #222; line-height: 1.5; }
    .top-bar { display: flex; justify-content: space-between; align-items: center; font-size: 14px; margin-bottom: 10px; }
    .contact span:not(:last-child)::after { content: ' \\2022 '; }
    .logo { height: 40px; }
    header { text-align: center; background: #990000; color: #fff; padding: 20px 0; margin-bottom: 20px; }
    h1 { font-size: 36px; margin: 0 0 12px; font-weight: 700; }
    h2 { font-size: 18px; margin: 30px 0 12px; background: #f2f2f2; padding: 6px 10px; color: #990000; font-weight: 700; text-transform: uppercase; }
    section { margin-bottom: 20px; }
    ul { list-style: none; padding-left: 0; margin: 0; }
    li {
      margin-bottom: 10px;
      margin-left: 1.2em;
      white-space: pre-wrap;
      line-height: 1.5;
      font-weight: 400;
    }
    .bullet {
      display: inline-block;
      width: 1em;
      margin-left: -1em;
      margin-right: 0.5em;
      color: #990000;
      font-weight: 700;
    }
    .edu-bullet {
      display: inline-block;
      width: 1em;
      margin-left: -1em;
      margin-right: 0.5em;
      color: #990000;
      font-weight: 700;
    }
    .tab { display:inline-block; width:1.5em; }
    .tab + p { margin-top: 0; }
    strong { font-weight: 700; }
  </style>
</head>
<body>
  <div class="top-bar">
    <div class="contact">
      <span>{{phone}}</span>
      <span>{{email}}</span>
      <span>{{cityState}}</span>
      <span>{{linkedin}}</span>
    </div>
    <img src="https://resumeforge.s3.amazonaws.com/ucmo-logo.png" alt="UCMO logo" class="logo" />
  </div>
  <header>
    <h1>Jane Doe</h1>
  </header>
  
  <section>
    <h2>Work Experience</h2>
    {{#if items}}
      <ul>
        {{#each items}}
          <li>{{{this}}}</li>
        
  <section>
    <h2>Education</h2>
    {{#if items}}
      <ul>
        {{#each items}}
          <li>{{{this}}}</li>
        
      </ul>
    {{/if}}
  </section>
  {{/each}}
</body>
</html>
"
`;
